[FLINK-7709] Add CheckpointStatisticDetailsHandler for new REST endpoint

tillrohrmann opened the pull request:
## What is the purpose of the change

Adds the new `CheckpointStatisticDetailsHandler` for the new REST server endpoint.

Moreover, this PR disables the `FAIL_ON_MISSING_CREATOR_PROPERTIES` property for the `RestMapperUtils.getStrictObjectMapper` because that is something the individuals beans can do on their own (e.g. by checking with `Preconditions.checkNotNull`).

R @zentol because of the changes to the `ObjectMapper` setup.

## Verifying this change

This change is already covered by existing tests, such as `CheckpointingStatisticsTest`.


## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

tillrohrmann commented:
Rebasing onto the latest master.

zentol commented:
Reviewed 24 of 31 files at r1.
Review status: 21 of 26 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/checkpoints/CheckpointStatsCache.java, line 44 at r1](https://reviewable.io:443/reviews/apache/flink/4763#-KvbgxjlkW5IZW5m8HYM:-KvbgxjlkW5IZW5m8HYN:bwa0bd4) ([raw file](https://github.com/apache/flink/blob/d152cf222422fd1565f4ea60fd6e8b62604a1dd1/flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/checkpoints/CheckpointStatsCache.java#L44)):*
> ```Java
> 	private final Cache<Long, AbstractCheckpointStats> cache;
> 
> 	public CheckpointStatsCache(int maxNumEntries) {
> ```

The CheckpointStatsCache should be moved out of the legacy namespace.

---

*[flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java, line 233 at r1](https://reviewable.io:443/reviews/apache/flink/4763#-Kvbj3-UxgqUuhmsfJcA:-Kvbj3-VGSGSiPo7duwE:b-ryryhy) ([raw file](https://github.com/apache/flink/blob/d152cf222422fd1565f4ea60fd6e8b62604a1dd1/flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java#L233)):*
> ```Java
> 				}
> 			} else {
> 				checkpointStatisticsPerTask = Collections.emptyMap();
> ```

This method is near identical to `CheckpointingStatistics#generate...`, except this one line.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/flink/4763)*
<!-- Sent from Reviewable.io -->

tillrohrmann commented:
@zentol I'm not entirely sure whether reviewable makes things better or worse wrt reviewing. At least I can no longer respond to individual comments directly from github.

You're right with both comments. I'll remove the `CheckpointingStatistics#generate` method. Moreover, I'll move the CheckpointStatsCache out of the legacy package.

zentol commented:
@tillrohrmann I wanted to try it out, primarily since i can mark individual files as reviewed. For the remaining files I will once again write the comments on github.

tillrohrmann commented:
Alright @zentol. I guess it would work if I signed up for reviewable.

tillrohrmann commented:
Quick question @zentol. You've reviewed and approved #4772 which is based on this one. Do you have more comments for this PR?

zentol commented:
I still have to look at the JSON generation in this PR.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -0,0 +1,534 @@
+			@JsonProperty(FIELD_NAME_NUM_SUBTASKS) int numSubtasks,
+			@JsonProperty(FIELD_NAME_NUM_ACK_SUBTASKS) int numAckSubtasks,
+			@JsonDeserialize(keyUsing = JobVertexIDDeserializer.class) @JsonProperty(FIELD_NAME_TASKS) @Nullable Map<JobVertexID, TaskCheckpointStatistics> checkpointStatisticsPerTask) {
+		this.id = id;
```
missing `checkNotNull` check for all fields but `checkpoitnStatisticsPerTask`.

tillrohrmann replied:
Good catch. Will add the checks.

tillrohrmann replied:
Hmm I think all not checked parameters are actually primitives and, thus, don't need to be checked for null.

zentol replied:
true, all objects are checked.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -0,0 +1,534 @@
+@JsonSubTypes({
+	@JsonSubTypes.Type(value = CheckpointStatistics.CompletedCheckpointStatistics.class, name = "completed"),
+	@JsonSubTypes.Type(value = CheckpointStatistics.FailedCheckpointStatistics.class, name = "failed")})
+@JsonInclude(JsonInclude.Include.NON_NULL)
```
We could annotate the particular field instead of the class.

tillrohrmann replied:
I will remove the Include.NON_NULL directive and re-add the FAIL_ON_MISSING_CREATOR_PROPERTY to the `ObjectMapper`.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -0,0 +1,534 @@
+				throw new IllegalArgumentException("Given checkpoint stats object of type " + checkpointStats.getClass().getName() + " cannot be converted.");
+			}
+		} else {
+			return null;
```
pretty sure this would lead to a NullPointerException, since neither the CheckpointStatshandler, not AbstractExceutionGraphhandler, nor the AbstractRestHandler handle the case of the returned value being null.

tillrohrmann replied:
I think it should actually work and output a serialized `null` value. However, I this distinction is not necessary and should be better pulled out of this method.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -0,0 +1,534 @@
+				numAckSubtasks,
+				checkpointingStatisticsPerTask);
+
+			this.externalPath = externalPath;
```
missing null checks

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java:
```diff
@@ -0,0 +1,534 @@
+				numAckSubtasks,
+				checkpointingStatisticsPerTask);
+
+			this.failureTimestamp = failureTimestamp;
```
missing null checks

tillrohrmann replied:
Should be a primitive.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/util/RestMapperUtils.java:
```diff
@@ -33,8 +33,7 @@
 			DeserializationFeature.FAIL_ON_IGNORED_PROPERTIES,
 			DeserializationFeature.FAIL_ON_NULL_FOR_PRIMITIVES,
-			DeserializationFeature.FAIL_ON_READING_DUP_TREE_KEY,
-			DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES);
```
Removing this also means we have to add explicit null checks to all existing Request-/ResponseBody classes.

zentol replied:
Do note that we must be on the look-out for requests that use primitive fields, as jackson will default them to 0 if they are missing, which will cause misleading error messages.

tillrohrmann replied:
This is a valid objection which I share. I'll remove this change and set the map of `TaskCheckpointStatistics` to an empty map in case that we want to leave the details out.

tillrohrmann commented:
Thanks for the review @zentol. I've addressed your comments: Reverted change to `RestMapperUtils`, removed null branch from `CheckpointStatistics#generateCheckpointStatistics`; Made the `CheckpointStatistics#checkpointStatisticsPerTask` non nullable. If you have no further objections, then I would like to merge this PR once Travis gives green light.

tillrohrmann commented:
Rebased onto the latest master

zentol commented:
no objections, feel free to merge this.
